# Сессионное задание по курсу «Математический анализ и теория вероятностей» Вдовина Д.Н

#### Задание 1.

Найдите производные функций. Подробно распишите решение. Задание должно быть решено без использования Python. 

(a). $y=e^{x\ln(3x+1)}$.

**Область определения.** Требуем $3x+1>0 \Rightarrow x>-1/3$.

Положим $g(x)=x\ln(3x+1)$. Тогда $y=e^{g(x)}$, и по правилу производной экспоненты
$$
y' = e^{g(x)}\cdot g'(x).
$$

Найдём $g'(x)$ (правило произведения):
$$
g'(x)=(x)'\ln(3x+1)+x\cdot\frac{(3x+1)'}{3x+1}
=\ln(3x+1)+\frac{3x}{3x+1}.
$$

**Итог:**
$$
\boxed{\,y' = e^{x\ln(3x+1)}\!\left(\ln(3x+1)+\frac{3x}{3x+1}\right),\quad x>-1/3\,}
$$





(б). $\,y=e^{x^2}\sqrt{x^3+4x^2-7}\,$.

**Область определения.** Подкоренное выражение неотрицательно:
$$
x^3+4x^2-7\;\ge 0.
$$

Положим $u(x)=e^{x^2}$ и $v(x)=\sqrt{x^3+4x^2-7}=(x^3+4x^2-7)^{1/2}$. Тогда $y=u\cdot v$, и по правилу произведения
$$
y' = u'\,v + u\,v'.
$$

Найдём производные $u'$ и $v'$ (правила цепочки и степени):
$$
u'(x)=\frac{d}{dx}\,e^{x^2}=2x\,e^{x^2},\qquad
v'(x)=\tfrac12\,(x^3+4x^2-7)^{-1/2}\cdot(3x^2+8x)=\frac{3x^2+8x}{2\sqrt{x^3+4x^2-7}}.
$$

Подставляя в формулу произведения, получаем
$$
y' \;=\; (2x\,e^{x^2})\sqrt{x^3+4x^2-7}\;+\;e^{x^2}\cdot\frac{3x^2+8x}{2\sqrt{x^3+4x^2-7}}.
$$

Вынесем общий множитель $e^{x^2}$ :
$$
\boxed{\;
y' \;=\; e^{x^2}\!\left( 2x\,\sqrt{x^3+4x^2-7}\;+\;\frac{3x^2+8x}{2\,\sqrt{x^3+4x^2-7}} \right),\quad
x^3+4x^2-7\ge 0.\;}
$$

Эквивалентная форма после приведения к общему знаменателю:
$\displaystyle
y'=\;e^{x^2}\,\frac{4x\,(x^3+4x^2-7)+(3x^2+8x)}{2\sqrt{x^3+4x^2-7}}
$.)


(в). 

$$
y=\frac{x\sqrt{1-x^2}}{1+x^2}.
$$

**Область определения.** Подкоренное выражение неотрицательно:
$$
1-x^2\ge 0 \;\Rightarrow\; |x|\le 1,\qquad 1+x^2>0.
$$
(Формула производной корректна при \(|x|<1\).)

Обозначим
$$
N(x)=x\sqrt{1-x^2},\qquad D(x)=1+x^2,\qquad y=\frac{N}{D}.
$$
Тогда по правилу частного
$$
y'=\frac{N'D-ND'}{D^{2}}.
$$

**Вычислим \(N'(x)\). Пусть**
$$
s(x)=\sqrt{1-x^2}=(1-x^2)^{1/2}.
$$
$$
s'(x)=\tfrac{1}{2}(1-x^2)^{-1/2}\cdot(-2x)=-\frac{x}{\sqrt{1-x^2}}.
$$
Отсюда
$$
N'(x)=1\cdot s(x)+x\cdot s'(x)
=\sqrt{1-x^2}-\frac{x^2}{\sqrt{1-x^2}}
=\frac{1-2x^2}{\sqrt{1-x^2}},\qquad
D'(x)=2x.
$$

Подставляя в формулу частного:
$$
y'=\frac{\dfrac{1-2x^2}{\sqrt{1-x^2}}\,(1+x^2)\;-\;x\sqrt{1-x^2}\cdot(2x)}{(1+x^2)^2}.
$$

Приведём к общему знаменателю и упростим:
$$
y'=\frac{(1-2x^2)(1+x^2)-2x^2(1-x^2)}{(1+x^2)^2\sqrt{1-x^2}}
=\frac{1-3x^2}{(1+x^2)^2\sqrt{1-x^2}}.
$$

**Итог**
$$
\boxed{\,y'=\dfrac{1-3x^2}{(1+x^2)^2\sqrt{1-x^2}}\,},\qquad |x|<1.
$$



### Задание 2

Найдите безусловные экстремумы функции:
Подробно распишите решение. Задание должно быть решено без использования Python.


$$
u = 3x^3+y^2+z^2+6xy-2z+1.
$$

**Шаг 1. Стационарные точки (градиент = 0).**

Частные производные:
$$
u_x=9x^2+6y,\qquad
u_y=2y+6x,\qquad
u_z=2z-2.
$$

Система $\,\nabla u=0\,$:

$$
\begin{cases}
9x^2+6y=0,\\
2y+6x=0,\\
2z-2=0
\end{cases}
\;\Rightarrow\;
\begin{cases}
y=-3x,\\
z=1,\\
9x^2-18x=0 \ \Rightarrow\ 9x(x-2)=0
\end{cases}
$$

Решаем последовательно:
\[
\begin{aligned}
&2z-2=0 \ &&\Rightarrow\ && z=1,\\
&2y+6x=0 \ &&\Rightarrow\ && y=-3x,\\
&9x^2+6(-3x)=9x^2-18x=9x(x-2)=0 \ &&\Rightarrow\ && x\in\{0,2\},\\
&y=-3x \ &&\Rightarrow\ && y\in\{0,-6\}.
\end{aligned}
\]

Следовательно, стационарные точки:
$$
A=(0,0,1),\qquad B=(2,-6,1).
$$

**Шаг 2. Классификация по гессиану.**

Вторые производные:
$$
u_{xx}=18x,\quad u_{yy}=2,\quad u_{zz}=2,\quad
u_{xy}=u_{yx}=6,\quad u_{xz}=u_{zx}=0,\quad u_{yz}=u_{zy}=0.
$$

Гессиан:
$$
H(x)=
\begin{bmatrix}
18x & 6 & 0\\
6   & 2 & 0\\
0   & 0 & 2
\end{bmatrix}.
$$

— В точке \(A\) ( \(x=0\) ):
$$
H(A)=
\begin{bmatrix}
0 & 6 & 0\\
6 & 2 & 0\\
0 & 0 & 2
\end{bmatrix},
\qquad
\det\!\begin{bmatrix}0&6\\[2pt]6&2\end{bmatrix}=-36<0.

$$
Гессиан неопределённый $ \Rightarrow $ $A$ — **седловая точка**.



— В точке \(B\) ( \(x=2\) ):
$$
H(B)=
\begin{bmatrix}
36 & 6 & 0\\
6  & 2 & 0\\
0  & 0 & 2
\end{bmatrix}.
$$
Главные миноры:
$$
\Delta_1=36>0,\qquad
\Delta_2=\det\!\begin{bmatrix}36&6\\[2pt]6&2\end{bmatrix}=36>0,\qquad
\Delta_3=\det H(B)=2\cdot 36=72>0.
$$
Гессиан положительно определён $ \Rightarrow $ $B$ — **локальный минимум**.

**Шаг 3. Значения функции в стационарных точках.**
$$
u(A)=3\cdot0^3+0^2+1^2+6\cdot0\cdot0-2\cdot1+1=0,
$$
$$
u(B)=3\cdot2^3+(-6)^2+1^2+6\cdot2\cdot(-6)-2\cdot1+1
=24+36+1-72-2+1=-12.
$$

**Вывод.**
$$
\boxed{\text{Локальный минимум: } B=(2,-6,1),\quad u_{\min}=-12.}
$$
$$
\boxed{\text{Точка } A=(0,0,1)\ \text{— седловая (экстремума нет).}}
$$

$$
\boxed{\text{Из-за члена \(3x^3\) функция не ограничена ни сверху, ни снизу; глобальных экстремумов нет.}}
$$


### Задание 3.
### Найдите условный экстремум функции, если задано условие:

Функция и ограничение

$$
z = e^{-(x+y)}, \qquad \frac{x^2}{4} + y^2 = 5 .
$$

---

### Метод множителей Лагранжа

Лагранжиан

$$
\mathcal{L}(x,y,\lambda) \;=\; e^{-(x+y)} \;+\; \lambda\!\left(\frac{x^2}{4}+y^2-5\right).
$$

Условия первого порядка

$$
\begin{cases}
\mathcal{L}_x = -\,e^{-(x+y)} + \lambda\,\dfrac{x}{2} = 0,\\[6pt]
\mathcal{L}_y = -\,e^{-(x+y)} + 2\lambda y = 0,\\[6pt]
\mathcal{L}_\lambda = \dfrac{x^2}{4} + y^2 - 5 = 0.
\end{cases}
$$

Из первых двух уравнений:

$$
\lambda\,\frac{x}{2} \;=\; 2\lambda y \quad \Rightarrow \quad x = 4y,
$$

$$\text{(случай }\lambda=0\text{ невозможен, так как } e^{-(x+y)}\neq 0\text{).}$$


Подстановка в ограничение:

$$
\frac{(4y)^2}{4}+y^2=5 \;\Rightarrow\; 4y^2+y^2=5 \;\Rightarrow\; y^2=1 \;\Rightarrow\; y=\pm 1,\; x=\pm 4.
$$

Кандидаты на условные экстремумы

$$
(4,1)\quad \text{и}\quad (-4,-1).
$$

---

### Классификация (минимум/максимум)

$$
z \;=\; e^{-(x+y)} \;\; \text{монотонно убывает по сумме } (x+y).
$$

$$
\frac{x^2}{4} + y^2 = 5 \;\; \text{— уравнение эллипса}.
$$

Максимальное значение \(x+y\) на эллипсе достигается при \((4,1)\), минимальное — при \((-4,-1)\). Поэтому

$$
(4,1)\; \Rightarrow\; x+y=5 \; \Rightarrow\; z=e^{-5}\ \text{ — условный минимум};
$$

$$
(-4,-1)\; \Rightarrow\; x+y=-5 \; \Rightarrow\; z=e^{5}\ \text{ — условный максимум}.
$$

---

### Ответ

$$
\text{Условный минимум:}\quad (x,y)=(4,1),\quad z_{\min}=e^{-5}.
$$

$$
\text{Условный максимум:}\quad (x,y)=(-4,-1),\quad z_{\max}=e^{5}.
$$


### Задание 4. 

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.naive_bayes import CategoricalNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# 1. Загрузка данных
data = pd.read_csv("exam.csv") # Загрузка данных из CSV файла
print(data.head())

# 2. Кодирование категориальных признаков
le_test = LabelEncoder() 
le_age = LabelEncoder() 
le_status = LabelEncoder()

data['Test'] = le_test.fit_transform(data['Test']) # Кодирование столбца 'Test'
data['Age_Group'] = le_age.fit_transform(data['Age_Group']) # Кодирование столбца 'Age_Group'
data['Status'] = le_status.fit_transform(data['Status']) # Кодирование целевого столбца 'Status'

# 3. Разделение данных на обучающую и тестовую выборки
X = data[['Test', 'Age_Group']] # Признаки
y = data['Status'] # Целевая переменная
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42, stratify=y
)

# 4. Обучение модели Наивного Байеса
model = CategoricalNB() # Инициализация модели
model.fit(X_train, y_train) # Обучение модели

# 5. Предсказание и оценка качества модели
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred) # Точность
precision = precision_score(y_test, y_pred) # Точность положительных предсказаний
recall = recall_score(y_test, y_pred) # Полнота
f1 = f1_score(y_test, y_pred) # F1-мера

print("Accuracy:", round(accuracy, 3)) # Вывод результатов точности
print("Precision:", round(precision, 3)) # Вывод результатов точности положительных предсказаний
print("Recall:", round(recall, 3)) # Вывод результатов полноты
print("F1-score:", round(f1, 3)) # Вывод результатов F1-меры


       Test Age_Group    Status
0  Positive     Young  Infected
1  Positive     Young  Infected
2  Positive     Young  Infected
3  Positive       Old  Infected
4  Positive       Old  Infected
Accuracy: 0.771
Precision: 0.725
Recall: 0.853
F1-score: 0.784


Модель корректно классифицирует случаи с точностью 77%.

Показатель recall = 0.853 говорит о том, что алгоритм хорошо выявляет заболевших (низкое число пропущенных).

Небольшое снижение precision до 0.725 означает, что часть здоровых людей могла быть ошибочно отнесена к заболевшим.

Итоговая метрика F1 = 0.784 показывает хорошее качество классификации.

### Задание 5.

In [4]:
# Метод Ньютона для оптимизации f(x) = x^4 + 3x^3 - 12x^2 + 7x - 2
# Старт: x0 = 10, точность: 1e-4

def f(x):
    return x**4 + 3*x**3 - 12*x**2 + 7*x - 2

def df(x):   # f'(x) первая производная
    return 4*x**3 + 9*x**2 - 24*x + 7

def d2f(x):  # f''(x) вторая производная
    return 12*x**2 + 18*x - 24

def newton_optimize(x0=10.0, tol=1e-4, max_iter=100): # Метод Ньютона
    x = float(x0) # начальная точка
    for it in range(1, max_iter + 1):
        g  = df(x) # градиент в текущей точке
        H  = d2f(x) # гессиан (вторая производная) в текущей точке
        if abs(H) < 1e-12:   # защита от деления на ~0 
            raise ZeroDivisionError("Вторая производная близка к нулю — метод Ньютона остановлен.")
        x_new = x - g / H # шаг метода Ньютона
        if abs(x_new - x) < tol or abs(g) < tol: # условие остановки
            x = x_new
            return x, f(x), it
        x = x_new
    return x, f(x), max_iter # возврат результата после max_iter

x_star, f_star, n_iter = newton_optimize(x0=10.0, tol=1e-4) # Запуск метода Ньютона

# Классификация (минимум/максимум) по знаку f''(x*)
kind = "минимум" if d2f(x_star) > 0 else ("максимум" if d2f(x_star) < 0 else "перелом/неопределено")

print(f"x* = {x_star:.6f}") # Вывод найденной точки x* где метод Ньютона остановился
print(f"f(x*) = {f_star:.6f}") # Вывод значения функции в точке x* 
print(f"итераций: {n_iter}") # Вывод количества итераций
print(f"тип точки: {kind}") # Вывод типа точки


x* = 1.310136
f(x*) = -3.733928
итераций: 9
тип точки: минимум


### Задание 6.

In [5]:
# Распределение: P(X=k) для k=0..6 заданы, P(X=7) неизвестна.
p = {0: 0.10, 1: 0.30, 2: 0.10, 3: 0.25, 4: 0.05, 5: 0.05, 6: 0.10}

# 1) Вероятность получить 7 писем
p7 = 1.0 - sum(p.values()) # P(X=7) = 1 - сумма остальных вероятностей
p[7] = p7  # добавим в распределение P(X=7)

# 2) Математическое ожидание E[X] 
EX = sum(k * pk for k, pk in p.items())# вычисление E[X]

# 3) Дисперсия Var[X] = E[X^2] - (E[X])^2
EX2 = sum((k * k) * pk for k, pk in p.items())
Var = EX2 - EX * EX 

# Вывод c округлением до сотых (требование задачи)
print(f"P(X=7) = {p7:.2f}") # Вывод вероятности P(X=7)
print(f"E[X] = {EX:.2f}") # Вывод математического ожидания E[X]
print(f"Var[X] = {Var:.2f}") # Вывод дисперсии Var[X]


P(X=7) = 0.05
E[X] = 2.65
Var[X] = 4.03


### Задание 7.

In [8]:
# Число звонков за час: N ~ Poisson(λ=2). Ищем P(2 ≤ N ≤ 4).
def fact(n):
    f = 1
    for i in range(2, n+1):
        f *= i
    return f

# простая экспонента 
import math
lam = 2.0
p = 0.0
for k in (2, 3, 4):
    p += math.exp(-lam) * (lam**k) / fact(k)

print("Искомая вероятность: P(2 ≤ N ≤ 4).")
print(f"Результат: P = {p:.3f}")

Искомая вероятность: P(2 ≤ N ≤ 4).
Результат: P = 0.541
